fileServer.js

```diff
@@ -0,0 +1,67 @@
+ 'use strict';
+
+ var http = require('http');
+ var fs = require('fs');
+ var path = require('path');
+ var url = require('url');
+ var rootDir = __dirname + '/public';
+
+ const mimeType = {
+   'html': 'text/html',
+   'jpeg': 'image/jpeg',
+   'jpg': 'image/jpeg',
+   'png': 'image/png',
+   'js': 'text/javascript',
+   'css': 'text/css'
+ };
+
+ function send404Request(res) {
+   fs.readFile('public/404.html', 'UTF-8', function(err, html) {
+     res.setHeader('Content-type', mimeType);
+     res.end(html);
+   });
+ }
+
+ function onRequest(req, res) {
+   console.log('request starting...');
+
+   var filename;
+   var pathname = url.parse(req.url).pathname;
+   filename = path.join(rootDir, pathname);
+   console.log('Filename: ' + filename);
+
+   if (req.method !== 'GET') {
+     res.writeHead(405, {
+       "Content-type": "text/plain"
+     });
+     res.write(`HTTP method ${req.method} not yet supported`);
+     res.end();
+     return
+   }
+
+   if (req.url === '/') {
+     res.writeHead(200, {
+       'Content-type': mimeType
+     });
+     fs.createReadStream('public/index.html').pipe(res);
+   }
+
+   fs.exists(filename, function(exists) {
+
+     if (exists) {
+       fs.readFile(filename, function(error, content) {
+         res.writeHead(200, {
+           'Content-type': mimeType
+         });
+         res.end(content, 'utf-8');
+
+       });
+     } else {
+       send404Request(res);
```

Comment:
IF the file does not exist you will never reach this because it is **inside the** `fs.exists` function, which is only called when the file does exist.

Also - you could get an error attempting to read the server side file with `fs.readFile` - make sure you handle the error and communicate to the operations person or developer that the read failed.

Reply:
This is still broken logic, Sandra.

Please read these pages for an overview of how permissions work for filesystems on Unix-based systems and a direct intro to filesystem tools used form the CLI :) - I will quiz you on the concepts and basic CLI  tools (ls, cp, ln, rm, cat, cd, less, more, head, tail, mkdir, rmdir, pwd) in our sessions together to make sure you understand them and how to use them.

http://www.ee.surrey.ac.uk/Teaching/Unix/

https://www.safaribooksonline.com/library/view/developing-bioinformatics-computer/1565926641/ch04.html

http://www.cis.rit.edu/class/simg211/unixintro/Filesystem.html